<launch>

  <arg name="x_pos" default="0.0"/>
  <arg name="y_pos" default="0.0"/>
  <arg name="yaw" default="1.5708"/>  

  <!-- Map server -->
  <node name="map_server" pkg="map_server" type="map_server" args="$(find tk8)/maps/turtle.yaml" output="screen">
    <param name="frame_id" value="map"/>
  </node>
  
  <!-- Static transform -->
  <node pkg="tf" type="static_transform_publisher" name="base_to_rotated_base" args="0 0 0 1.5708 0 0 base_link rotated_base_link 100" />
  
  <!-- AMCL for localization -->
  <node pkg="amcl" type="amcl" name="amcl" output="screen">
    <!-- AMCL parameters -->
    <param name="odom_frame_id" value="odom"/>
    <param name="base_frame_id" value="rotated_base_link"/>    
    <param name="global_frame_id" value="map"/>
    
    <!-- Initial Pose -->
    <param name="initial_pose_x" value="$(arg x_pos)"/>
    <param name="initial_pose_y" value="$(arg y_pos)"/>
    <param name="initial_pose_a" value="$(arg yaw)"/>  <!-- Yaw = Z-rotation -->
    
    <!-- Laser model parameters -->
    <param name="laser_max_beams" value="30"/>
    <param name="laser_z_hit" value="0.95"/>
    <param name="laser_z_short" value="0.1"/>
    <param name="laser_z_max" value="0.05"/>
    <param name="laser_z_rand" value="0.05"/>
    <param name="laser_sigma_hit" value="0.2"/>
    <param name="laser_lambda_short" value="0.1"/>
    <param name="laser_model_type" value="likelihood_field"/>
    
    <!-- Update parameters -->
    <param name="update_min_d" value="0.2"/>
    <param name="update_min_a" value="0.5"/>
    <param name="resample_interval" value="1"/>
    
    <!-- Other parameters -->
    <param name="recovery_alpha_slow" value="0.0"/>
    <param name="recovery_alpha_fast" value="0.0"/>
  </node>

  <!-- Move base -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <!-- Common parameters -->
    <param name="controller_frequency" value="10.0"/>
    <param name="controller_patience" value="3.0"/>
    <param name="recovery_behavior_enabled" value="true"/>
    <param name="clearing_rotation_allowed" value="true"/>
    <param name="oscillation_timeout" value="10.0"/>
    <param name="oscillation_distance" value="0.2"/>
    
    <!-- Global and local planners -->
    <param name="base_global_planner" value="navfn/NavfnROS"/>
    <param name="base_local_planner" value="base_local_planner/TrajectoryPlannerROS"/>
    
    <!-- Topics -->
    <remap from="cmd_vel" to="cmd_vel"/>
    <remap from="odom" to="odom"/>
    <remap from="scan" to="scan"/>

    <!-- Global costmap parameters -->
    <rosparam file="$(find tk8)/config/costmap_common_params.yaml" command="load" ns="global_costmap"/>
    <rosparam file="$(find tk8)/config/global_costmap_params.yaml" command="load"/>
    
    <!-- Local costmap parameters -->
    <rosparam file="$(find tk8)/config/costmap_common_params.yaml" command="load" ns="local_costmap"/>
    <rosparam file="$(find tk8)/config/local_costmap_params.yaml" command="load"/>
    
    <!-- Trajectory planner parameters -->
    <rosparam file="$(find tk8)/config/base_local_planner_params.yaml" command="load"/>
  </node>

  <!-- RViz for visualization -->
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find tk8)/tk8_default.rviz"/>
</launch>
